### PR TITLE
[WB-4543] Add nested metric support to run.define_metric()

### PR DESCRIPTION
**Pull request by raubitsj**
https://wandb.atlassian.net/browse/WB-4543

Description
-----------

Finish nested define_metric support from previous PR:
https://github.com/wandb/client/pull/1802

Design
-------
https://paper.dropbox.com/doc/CLI-API-add-run.define_metric--BEm98DroFga614g7tvrdo6IXAg-z9GIYShHF1sa9fTlBfxxz
(this design has evolved over time but it captures many of the design decisions which were considered)
This PR serves as the most up-to-date information about the change.

Proposed changes based on feedback
---------------------------------
- Support nested histories, '.' used as separator.  If the user has "."s in their metric they will have to escape them. See examples
- Support nested aggregations.  If specified, the leaf value (which is currently the last_value) becomes a dictionary.  See examples.
- Add "last" aggregation, if enabled it will change summary
from:
{"metric":{"is":{"nested": {"min": 3}}}
to:
{"metric":{"is":{"nested": {"min": {"last": 3}}}}
- "mean" aggregation should also log "num" or "count" to eventually support resume
- add more validation to define_metric name string.  check for invalid chars, etc
- add "none" to summary, so we can avoid even capturing it for some metrics (need to clear others when seen, not siimple merge
- add user intent to json metadata,  previously the code would not inform the frontend of the glob, instead it would match metrics as they were seen,  While this is useful to keep frontend more simple. it hides information that might be useful.  A key case is determining x-axis on grouped runs.  If the last run is used to find the preferred x-axis, and this run ended early without logging the metric. then the frontend wont know to apply the x-axis.  The "fix" though complicated, is so the frontend could know the intent was to glob. and would have to apply a similar glob logic, to decide how to handle this grouped view.   Any user intent record will have a flag indicating that they are intent.

Only required change is to make sure the metadata has a name that is properly escaped.  everything else is optional for now as this PR doesnt support nested histories (yet)

Examples
----------

1) Metric with dots
Metric name: "this.metric.has.dots"
```
wandb.define_metric("this\.metric\.has\.dots", ...)
wandb.define_metric("this.metric.has.dots", ...).  <== Possibly allow (LATER)
wandb.log({"this.metric.has.dots": 3})
```
aggregations will be:
```
"this.metric.has.dots.min"
"this.metric.has.dots.max"
```

metric in metadata will be: (MUST be escaped)
```
"m": [ {"1": "this\.metric\.has\.dots",
```

Note 1:
It would be good to have a warning at the end of the run if a metric did not match, this would indicate a problem to the user.
Note 2:
We might accept the non-escaped string and match both for nested dict and the flat string.   REEQUIREMENT: the actual metadata needs to be escaped properly -- this makes sure that there is no ambiguity in the data stored.   The overhead of this double check is low and it simplifies the user experience. only risk is that we will match a flat string even if the user only wanted the dict form to be matched, this seems like a very small risk. 

2) Nested Metric
```
wandb.define_metric("metric.is.nested, ...)
wandb.log(dict(metric=dict(is=dict(nested=3))))
```
when aggregations are not set summary will remain:
```
{"metric":{"is":{"nested": LAST}}}
```

when aggregations are set:
- {"metric":{"is":{"nested": {"min": 3}}}
-  {"metric":{"is":{"nested": {"last": 3}}}

metric in metadata will be:
```
"m": [ {"1": "metric.is.nested",
```

3) mixed
```
wandb.define_metric("metric.is\.nested, ...)
wandb.define_metric("metric.is.nested, ...).  <== Possibly allow (LATER)
wandb.log({"metric":{"is.nested": 3}})
```

metric in metadata will be:
```
"m": [ {"1": "metric.is\.nested",
```

Highlights
----------
- Aggregations like (min,max,mean,best) can be automatically added to the summary
- Integrations can specified their preferred step metric (and pass that info to the UI) eg "global_step" instead of _step
- wandb.log() code can better coexist with integration logged data by using the `step_sync` option (configured automatically by the integration (See: https://github.com/PyTorchLightning/pytorch-lightning/pull/5931)
- independent-variable metrics can be hidden from auto charts (`hide`)
- user now has a way to provide additional information about a metric (including optimization direction - to be used for auto sweeps).   But we could do more here like tag the metric with some metadata that this is a loss curve, etc which could help match the data with custom charts.  could also enforce types, or specify limitations (non-decreasing values)

Open Questions
----------------
- Is the UI / backend team ok with this protocol buffer json encoding (already being used for telemetry data).  It would be better for this to be structured data but synchronizing backend and cli changes are hard so this tries to create a format that is small, but fully forward/backward compatible.   Currently it is in the _wandb field of config, but it should go into the gorilla metadata field (when that is available).   If the backend team has a better proposal, it is very easy to change CLI to store it some other way.

Testing
-------

- unit tests 100% covered diff
- regression tests passing
- Prototype UI changes https://github.com/wandb/core/pull/6460
- PytorchLightning work
   - https://github.com/PyTorchLightning/pytorch-lightning/pull/5931
   - https://colab.research.google.com/drive/1W8Rg4ZEjISNPX8q7jyyQaltku7uwFccv?usp=sharing

TODO
------
- [ ] Test with new UI support

Example
--------
```python
import math
import wandb
run = wandb.init(project="WB-4543")
run.define_metric("global_step", hide=True)
run.define_metric("*", step_metric="global_step", step_sync=True, summary="min,max,best,mean", goal="maximize")
for x in range(5, 100):
    run.log(dict(global_step=x))
    val = 100*math.exp(-(((x-30)/15)**2)) + 60*math.exp(-(((x-70)/20)**2))
    run.log(dict(val=val))
```

Output
-------
Screen Shot 2021-02-15 at 3 50 44 PM

Screen Shot 2021-02-15 at 3 55 11 PM

- https://wandb.ai/jeffr/WB-4543/runs/5j451wv0
- https://dashboard.featurepeek.com/peek/0jvhnmv4#/jeffr/WB-4543/runs/5j451wv0

Metric data sent to UI:
```json
{
 "_wandb": {
        "desc": null,
        "value": {
            "m": [
                {
                    "1": "global_step",
                    "6": [
                        2
                    ]
                },
                {
                    "1": "val",
                    "5": 1,
                    "6": [
                        1
                    ],
                    "7": [
                        1,
                        2,
                        3,
                        4
                    ],
                    "8": 2
                }
            ],
      }
 }
}
```
Sample function to decode x-axis for a given metric:
https://github.com/wandb/client/blob/feature/define-metric/tests/utils/parse_metrics.py

```python
def get_step_metric_dict(ml):
    """Get mapping from metric to preferred x-axis."""
    nl = [m["1"] for m in ml]
    md = {m["1"]: nl[m["5"] - 1] for m in ml if m.get("5")}
    return md
```

### CODE REVIEW

**Review by vanpelt — COMMENTED** [review]
TDD!  😎

File: tests/integrations/test_keras.py [hunk 1]
```diff
@@ -80,6 +80,9 @@ def test_basic_keras(dummy_model, dummy_data, wandb_init_run):
     assert len(graph_json(wandb.run)["nodes"]) == 3
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 5), reason="test is flakey with py2, ignore for now"
+)
```
Comment (KyleGoyette) [inline tests/integrations/test_keras.py:85]: Thanks for doing this

File: wandb/sdk_py27/internal/handler.py [hunk 2]
```diff
@@ -178,81 +197,127 @@ def _save_history(self, record):
     def _update_summary_metrics(
         self,
         s,
-        k,
+        kl,
         v,
         float_v,
         goal_max,
     ):
+        print("UPDSUMMET", kl, v, s)
```
Comment (KyleGoyette) [inline wandb/sdk_py27/internal/handler.py:205]: Forgot to remove?

**Review by KyleGoyette — APPROVED** [review]
Nice, there's one loose print that I think we should pull out in handler

File: wandb/sdk_py27/internal/handler.py (hunk 2 above)
Comment (raubitsj) [reply wandb/sdk_py27/internal/handler.py:205]: oops. fixed.
